### DISCUSSION TITLE
Bi-directional relations

### DISCUSSION THREAD

**Discussion by ndarilek** [pocketbase/pocketbase#4493] [Q&A] [answered]
This seems like it should be simple but I'm struggling a bit with it and am wondering how to model it correctly. In my app, users add to an images collection which, on create, sends the image to an LLM for description and generates responses in the followups collection. In my current schema, users have many images, each of which has many followups.

Now I need API access rules. Users can only access images they've created. Likewise, users can only access followups for images they've created.

Looking at the schema in the database, though, I don't see a way to work backwards from images to users. It looks like maybe I can use `?=` to work forward in API rules (E.g. `id ?= @request.auth.images` to ensure that a user can only list their own images) but I'm not clear how I can do that with followups.

Do I need to create a has-one relationship on followups to users, or is there a way to work backward? Or do I need to create and manually maintain each backward relation?

Thanks.

**Comment by ganigeorgiev** [comment]
_If you haven't already you could check the [Back-relations docs](https://pocketbase.io/docs/working-with-relations/#back-relations)._

I'm not fully sure that I understand your data structure but if we make the following assumptions:
- the `users` collection has multiple `images` relation field pointing to the `images` collection
- the `images` collection has multiple `followups` relation field pointing to the `followups` collection

Then an example list or view API rules could be:

- for the `images` collection:
   ```js
   @request.auth.id != "" &&
   @request.auth.images.id ?= id
   ```

- for the `followups` collection:
   ```js
   @request.auth.id != "" &&
   @request.auth.images.followups.id ?= id
   ```

The above is just an example and will depend on your actual data structure and the desired data flow (for example, if a followup can be assigned to only 1 image, instead of multiple `relation` field in the `images` collection, you could use a single `relation` field in the `followups` collections pointing to its `images` record).

If the above doesn't help, please elaborate on your collections structure and desired behavior (you can also export your collections configuration as JSON from the _Admin UI > Settings > Export collections_).

_Edit: updated the `followups` rule_

**Reply by ndarilek** [reply]
Cool, thanks. Getting the following when attempting to add to images when not authenticated:

POST /api/collections/images/records
status: 400execTime: 14886.059802msauth: guestuserIp: 70.113.130.102error: Failed to create record.details: DrySubmit create rule failure: sql: no rows in result set

So good that it failed, but I'd have expected HTTP 403 which is what I think I got before. Is this the error I should expect for no user being authenticated with this particular rule?

Here are my collections. I tried adding API rules for images but am not sure how to add API rules for followups. Basically this is an image description service where each image has followups either created by the describer, or as questions about the image from the user. Images have multiple followups and followups are scoped to the image. Here are my current collections, thanks for all the help so far:

```
[
    {
        "id": "_pb_users_auth_",
        "name": "users",
        "type": "auth",
        "system": false,
        "schema": [
            {
                "system": false,
                "id": "users_name",
                "name": "name",
                "type": "text",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "min": null,
                    "max": null,
                    "pattern": ""
                }
            },
            {
                "system": false,
                "id": "users_avatar",
                "name": "avatar",
                "type": "file",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "mimeTypes": [
                        "image/jpeg",
                        "image/png",
                        "image/svg+xml",
                        "image/gif",
                        "image/webp"
                    ],
                    "thumbs": null,
                    "maxSelect": 1,
                    "maxSize": 5242880,
                    "protected": false
                }
            },
            {
                "system": false,
                "id": "fuov38bp",
                "name": "images",
                "type": "relation",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "collectionId": "reqb401a7t4vy65",
                    "cascadeDelete": false,
                    "minSelect": null,
                    "maxSelect": null,
                    "displayFields": null
                }
            }
        ],
        "indexes": [],
        "listRule": "id = @request.auth.id",
        "viewRule": "id = @request.auth.id",
        "createRule": "",
        "updateRule": "id = @request.auth.id",
        "deleteRule": "id = @request.auth.id",
        "options": {
            "allowEmailAuth": true,
            "allowOAuth2Auth": true,
            "allowUsernameAuth": true,
            "exceptEmailDomains": null,
            "manageRule": null,
            "minPasswordLength": 8,
            "onlyEmailDomains": null,
            "onlyVerified": false,
            "requireEmail": false
        }
    },
    {
        "id": "af4w81ckjzsuayr",
        "name": "followups",
        "type": "base",
        "system": false,
        "schema": [
            {
                "system": false,
                "id": "xeh717ou",
                "name": "user",
                "type": "bool",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {}
            },
            {
                "system": false,
                "id": "pqetxnfw",
                "name": "text",
                "type": "text",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "min": null,
                    "max": null,
                    "pattern": ""
                }
            }
        ],
        "indexes": [],
        "listRule": null,
        "viewRule": null,
        "createRule": null,
        "updateRule": null,
        "deleteRule": null,
        "options": {}
    },
    {
        "id": "reqb401a7t4vy65",
        "name": "images",
        "type": "base",
        "system": false,
        "schema": [
            {
                "system": false,
                "id": "jiadtasn",
                "name": "file",
                "type": "file",
                "required": true,
                "presentable": false,
                "unique": false,
                "options": {
                    "mimeTypes": [
                        "image/jpeg",
                        "image/png",
                        "image/svg+xml",
                        "image/gif",
                        "image/webp"
                    ],
                    "thumbs": [],
                    "maxSelect": 1,
                    "maxSize": 20000000,
                    "protected": true
                }
            },
            {
                "system": false,
                "id": "5y7zrgyt",
                "name": "followups",
                "type": "relation",
                "required": false,
                "presentable": false,
                "unique": false,
                "options": {
                    "collectionId": "af4w81ckjzsuayr",
                    "cascadeDelete": true,
                    "minSelect": null,
                    "maxSelect": null,
                    "displayFields": null
                }
            }
        ],
        "indexes": [],
        "listRule": "@request.auth.id != \"\" &&\n@request.auth.images.id ?= id",
        "viewRule": "@request.auth.id != \"\" &&\n@request.auth.images.id ?= id",
        "createRule": "@request.auth.id != \"\"",
        "updateRule": null,
        "deleteRule": null,
        "options": {}
    }
]
```

**Reply by ganigeorgiev** [reply]
> So good that it failed, but I'd have expected HTTP 403 which is what I think I got before. Is this the error I should expect for no user being authenticated with this particular rule?

This is working as expected. From the [docs](https://pocketbase.io/docs/api-rules-and-filters/):

> PocketBase API Rules act also as records filter!
> Or in other words, you could for example allow listing only the "active" records of your collection, by using a simple filter expression such as: status = "active" (where "status" is a field defined in your Collection).
>
> Because of the above, the API will return 200 empty items response in case a request doesn't satisfy a listRule, 400 for unsatisfied createRule and 404 for unsatisfied viewRule, updateRule and deleteRule.
All rules will return 403 in case they were "locked" (aka. admin only) and the request client is not an admin.

**Reply by ndarilek** [reply]
Ugh, sorry I missed that. I've been reading lots of docs and they're thorough, I just don't recall everything I read. :)

Given the collection structure above, how would I limit users to listing/viewing only folowups for images they've created? Guessing I have to add a relation from followups to their image and manually manage that? Followups are only created programatically so that should be doable.

Thanks again.

**Reply by ganigeorgiev** [reply]
> how would I limit users to listing/viewing only folowups for images they've created?

Based on your collections structure this rule should suffice

```js
@request.auth.id != "" &&
@request.auth.images.followups.id ?= id
```

~If it doesn't work make sure that you use the latest PocketBase version (back-relation filter and rules support was added in PocketBase v0.22.0).~ _You have direct relation from users->images so back-relation is not needed here._

**Reply by ganigeorgiev** [reply]
_Edit: Sorry, updated the rule as I forgot that the `user` has direct relation to the `images`._

**Comment by ndarilek** [comment]
Sorry, this is a work project and I was out of office last week.

Thanks for all this help. I have a couple more questions which I hope should clear things up:

1. The API rule in your last opost:
```
@request.auth.id != "" &&
@request.auth.images.followups.id ?= id
```
Where should I put that rule? It's not immediately obvious if it should go on `users`, `images` or `followups`.
2. I'm a bit confused about what happens to relations fields when referenced in API rules. In particular, `images.followups`... Does that essentially loop over every image in `images` and check `followups` for containment with `?=`? If so, are there any limitations to that? E.g. does that work over several levels of relations and expect the final bit of the rule to either resolve against a single value or to test containment? Sorry if I missed that in the docs--I looked specifically for that answer and didn't find a lot of clarity.

Aside: I wonder if a syntax like `@request.auth.images[].followups.id ?= id` might make it a bit more clear that this is possible?

Thanks.

**Reply by ganigeorgiev** [reply, marked as answer]
1. To avoid confusion, here are again the suggested List and View API rules:

    - for the `images` collection:
        ```js
        @request.auth.id != "" &&
        @request.auth.images.id ?= id
        ```

    - for the `followups` collection:
        ```js
        @request.auth.id != "" &&
        @request.auth.images.followups.id ?= id
        ```

2. The easiest way to think about the API rules is if they were part of a `WHERE` SQL statement, because that is what actually is happening under the hood.

    When you use dot-notation to target a relation field, we perform a `JOIN` with the related collection(s) and apply the condition against the last part of the expression.
    
    For example the last `followups` API rule generates a pseudo-query like:
    ```sql
    SELECT DISTINCT followups.*
    FROM followups
    LEFT JOIN users ON users.id = {:authId}
    LEFT JOIN images as usersImages ON users.images HAS usersImages.id -- fake HAS operator to simplify the example
    LEFT JOIN followups as imagesFollowups ON usersImages.followups HAS imagesFollowups.id
    WHERE (
        users.id != "" AND
        imagesFollowups.id = followups.id
    )
    ```
    Or the rule roughly translates to:
    _Require the user to be authenticated and **at least one** of its "images" records to have **at least one** "followups" record matching the id of the followup we are targeting._

    If in our rule the operator was only `=` and not `?=` in this case since the expression contains one or more multiple relations (_in our example that are the `images` and `followups` relation fields_) it will generate "match all" query, aka. query similar to the above one but with extra subquery(ies) that check that all related "images" and all of their "followups" records satisfy the constraint.

    There is no support at the moment for mixed (both "match-all" and "any/at-least-one") type of evaluation for a single nested relations expression.

    Nested relations are supported up to 6-levels depth, counting only the relation fields (if the limit is reached it returns an error). In the above example the nested relations depth is 2 - `images.followups`.


The documentation definitely needs more work but I'm not sure how to explain the rules better without complicating it too much with internal details.
I'll have to think a little more on this after the refactoring.

**Comment by ndarilek** [comment]
Thanks a bunch for all this help!
